rosplan_action_interface/src/RPKnowledgeBaseLink.py

```diff
@@ -0,0 +1,241 @@
+        bound_parameters = {}
+        for op_parameter in operator_details.formula.typed_parameters:
+            found = False
+            for i in range(0, len(pddl_action_msg.parameters)):
+                if op_parameter.key == pddl_action_msg.parameters[i].key:
+                    bound_parameters[pddl_action_msg.parameters[i].key] = pddl_action_msg.parameters[i].value
+                    found = True
+                    break
+            if not found:
+                rospy.logerr('KCL: ({}) aborting applying action effect due to missing parameter, missing {}'.format(rospy.get_name(), op_parameter.key))
+                return
+
+        kus.knowledge = []
+        kus.update_type = []
+        if effect_time == self.AT_START:
+            # At start add effects
+            for eff in operator_details.at_start_add_effects:
+                self.pack_simple_effect(eff, bound_parameters, kus, add_effect=False)
+            # At start del effects
+            for eff in operator_details.at_start_del_effects:
+                self.pack_simple_effect(eff, bound_parameters, kus, add_effect=True)
+
+        if effect_time == self.AT_END:
+            # At end add effects
+            for eff in operator_details.at_end_add_effects:
+                self.pack_simple_effect(eff, bound_parameters, kus, add_effect=False)
+            # At end del effects
+            for eff in operator_details.at_end_del_effects:
+                self.pack_simple_effect(eff, bound_parameters, kus, add_effect=True)
+
+        self.update_kb(kus)
+
+    def ground_simple_effect(self, effect, bound_parameters):
+
+        # TODO fetch predicate labels
+
+        # bind objects from action message to effect predicate
+        item = KnowledgeItem()
+        item.knowledge_type = KnowledgeItem.FACT
+        item.attribute_name = effect.name
+        for j in range(0, len(effect.typed_parameters)):
+            # set key (parameter label in predicate) and value (object bound to operator)
+            pair = KeyValue()
+            pair.key = "TODO THE LABEL"# self.predicates[self.op.at_start_del_effects[i].name].typed_parameters[j].key
+            pair.value = bound_parameters[effect.typed_parameters[j].key]
+            item.values.append(pair)
+
+        return item
+
+    # ================================== #
+    # update service(s) to knowldge base #
+    # ================================== #
+        
+    def update_kb(self, knowledge_update_service_request):
+        if len(knowledge_update_service_request.update_type) > 0:
+            try:
+                self.update_kb_srv.call(knowledge_update_service_request)
+            except Exception as e:
+                rospy.logerr("KCL Failed to update knowledge base: %s" % e.message)
+
+    # =============================== #
+    # request data from knowldge base #
+    # =============================== #
+
+    def get_kb_attribute(self, attribute_name):
+        # TODO error catching
```

Comment:
What kind of error catching would you like to add here and in the functions below?

Reply:
ServiceExcpetion failure for one. Adding.

Reply:
Added.